### PR TITLE
GCN Node Classification

### PR DESCRIPTION
**Pull request by Xiangxu-66**
Name: Xiangxu Zhang
Student Number: 47643439
This request implements a solution for multi-class node classification for Facebook network datasets using Graph Convolutional Networks (GCN). Included are the model built and a README, which documents the task, model, algorithm, dataset, usage, and results of the implementation.
Files Added:
- `dataset.py`  This file handles loading and processing the Facebook Large Page-Page Network dataset for a Graph Neural Network (GNN).
- `modules.py` This file uses PyTorch to define a multi-layer graph convolutional network (GCN) for semi-supervised multi-class node classification.
- `train.py` This file is used to handle the training of multi-layer graph neural networks (GNN). 
- `predict.py` This file is used to evaluate the trained GNN, computing the accuracy on the validation and test sets. Finally, t-SNE visualized node embedding is used to compress the classification results of the model into a two-dimensional space.
- README.md: Report the GCN model and its usage instructions.
- ./plots: Images of the generated plots in training and t-SNE visualization.

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**
GNN FB Page-Page Network dataset ---- Normal Difficulty
| Category	| | Marks | Comments |	
| :---  | :---: | :---: | :---  |		
| Algorithm solves the problem 		| 5 | 5 | |
| Implementation functions as intended	| 3 | 3 | 4 layer GCN, torch_geometric/GCNConv |
| Good design 					| 1 | 1 | Modular, Reusable |
| Commenting 					| 1 | 1 | Meaningful docstrings and comments |
| Algorithm above Normal Difficulty 	| 5 | 5 | |
| Algorithm is Hard difficulty 			| 5 | 0 | Normal Difficulty |
| Section IV : Max mark 15 from 20		|   | **15**   | |

**TSNE / UMAP:** 
Good
- Discussion: Good, with a bit more depth useful.

**Suggestions:**
Setup/Reproducibility/Requirements would be good to have.
Could add a hyper parameter search or discuss how the values were obtained.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. GNN Task.</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. Minimal.</td><td>-1</td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Acceptable. Usage info missing.</td><td>-2</td></tr>
	<tr><td></td><td>Description and Comments :Good. 11</td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is adequate. Add some info about the specific implementation here.</td><td>-1</td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-9</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_
